src/Microsoft.AspNet.SignalR.SqlServer/DbOperation.cs

```diff
@@ -180,43 +168,28 @@ private void LoggerCommand(DbCommand command)
         [SuppressMessage("Microsoft.Usage", "CA2202:Do not dispose objects multiple times", Justification = "Disposed in async Finally block"),
          SuppressMessage("Microsoft.Reliability", "CA2000:Dispose objects before losing scope", Justification = "Disposed in async Finally block")]
 #if ASPNET50
-        private void Execute<T>(Func<IDbCommand, Task<T>> commandFunc, TaskCompletionSource<T> tcs)
+        private async void Execute<T>(Func<IDbCommand, Task<T>> commandFunc, TaskCompletionSource<T> tcs)
 #else
-        private void Execute<T>(Func<DbCommand, Task<T>> commandFunc, TaskCompletionSource<T> tcs)
+        private async void Execute<T>(Func<DbCommand, Task<T>> commandFunc, TaskCompletionSource<T> tcs)
```

Comment:
Yes, @moozzyk is right, if you want to convert this pattern, you can do the following things:

1.
```C#
void DoSomethindAsync(Func<Task<T>> callback, TaskCompletionSource<T> tcs)
{
   callback().ContinueWith(async t =>
   {
       try 
       {
          await t;  // Await here so that we can observe any exceptions
          tcs.TrySetResult(t.Result);
       }
       catch (TaskCancelledException)
       {
           tcs.TrySetCancelled();
       }
       catch (Exception ex)
       {
           tcs.SetUnwrappedException(ex);
       }
   });
}
```